flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1394,13 +1394,14 @@ private CompletableFuture<Void> waitForTerminatingJob(
                                                     throwable));
                                 });
 
-        return jobManagerTerminationFuture.thenAcceptAsync(
+        return FutureUtils.thenAcceptAsyncIfNotDone(
```

Comment:
There is also a chance to submit duplicates when jobManagerTerminationFuture is not finished. Maybe we should introduce something like "Set\<JobID\> submitAndWaitingTermiateJobs". The Job that JobManagerTerminationFuture is not finished should be added, and deleted when the future is finished. Then we should check it in Dispatcher#isDuplicateJob.

Reply:
Thanks for the suggestion @huwh.

I added a new `submittedAndWaitingTerminationJobs` set for the outstanding jobIds, however, I wasn't sure if the lifecycle of the items in there need to be following the lifecycle of items in `jobManagerRunnerTerminationFutures` in any way.